{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e3fed514",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import yaml\n",
    "\n",
    "# from tsv2bedRMod import bid_mouse2bedRMod, bid_human2bedRMod, etam2bedRMod, csv2bedRMod\n",
    "from tsv2bedRMod import csv2bedRMod\n",
    "\n",
    "from helper import write_header\n",
    "from helper import get_modification_color\n",
    "from helper import parse_excel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f6dee60b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Table Legends & Header Info', '1. Candidate m1A sites', '2. All A sites with mut(hg19)', '3. All A sites with mut(rep)'])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keys = parse_excel(\"/home/annebusch/anne02/euf-data/rbs-seq/GSE90963_Table_S3-m1A_candidate_sites.xlsx\")\n",
    "keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67bbdedf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coordinate</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Base</th>\n",
       "      <th>BS_Depth</th>\n",
       "      <th>BS_A</th>\n",
       "      <th>BS_C</th>\n",
       "      <th>BS_G</th>\n",
       "      <th>...</th>\n",
       "      <th>Fold (NBS/BS).1</th>\n",
       "      <th>NBS_Depth</th>\n",
       "      <th>NBS_A</th>\n",
       "      <th>NBS_C</th>\n",
       "      <th>NBS_G</th>\n",
       "      <th>NBS_T</th>\n",
       "      <th>NBS_Mut</th>\n",
       "      <th>NBS_%C</th>\n",
       "      <th>NBS_%G</th>\n",
       "      <th>NBS_%T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_145822.1:12</td>\n",
       "      <td>.</td>\n",
       "      <td>28S</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>816923</td>\n",
       "      <td>813695</td>\n",
       "      <td>733</td>\n",
       "      <td>1487</td>\n",
       "      <td>...</td>\n",
       "      <td>2.990034</td>\n",
       "      <td>1023372</td>\n",
       "      <td>1016782</td>\n",
       "      <td>1992</td>\n",
       "      <td>3379</td>\n",
       "      <td>1219</td>\n",
       "      <td>0.006439</td>\n",
       "      <td>0.001947</td>\n",
       "      <td>0.003302</td>\n",
       "      <td>0.001191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_145822.1:1322</td>\n",
       "      <td>.</td>\n",
       "      <td>28S</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>197730</td>\n",
       "      <td>188581</td>\n",
       "      <td>102</td>\n",
       "      <td>2959</td>\n",
       "      <td>...</td>\n",
       "      <td>8.150468</td>\n",
       "      <td>26665</td>\n",
       "      <td>14725</td>\n",
       "      <td>88</td>\n",
       "      <td>4280</td>\n",
       "      <td>7572</td>\n",
       "      <td>0.447778</td>\n",
       "      <td>0.003300</td>\n",
       "      <td>0.160510</td>\n",
       "      <td>0.283968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tRNA_11:60</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_nmt-tRNA-Leu-TAA-1-1 (tRNAscan-SE...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>387</td>\n",
       "      <td>386</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>201.656000</td>\n",
       "      <td>625</td>\n",
       "      <td>170</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>443</td>\n",
       "      <td>0.728000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019200</td>\n",
       "      <td>0.708800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tRNA_41:36</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_tRNA-Ala-AGC-4-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>1155</td>\n",
       "      <td>1139</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>13.694228</td>\n",
       "      <td>641</td>\n",
       "      <td>489</td>\n",
       "      <td>8</td>\n",
       "      <td>51</td>\n",
       "      <td>93</td>\n",
       "      <td>0.237129</td>\n",
       "      <td>0.012480</td>\n",
       "      <td>0.079563</td>\n",
       "      <td>0.145086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tRNA_48:36</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_tRNA-Ala-CGC-1-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>481</td>\n",
       "      <td>462</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>50.517572</td>\n",
       "      <td>313</td>\n",
       "      <td>179</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>92</td>\n",
       "      <td>0.428115</td>\n",
       "      <td>0.009585</td>\n",
       "      <td>0.124601</td>\n",
       "      <td>0.293930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>tRNA_434:58</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_tRNA-Val-TAC-1-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>37244</td>\n",
       "      <td>37182</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>5.416748</td>\n",
       "      <td>7556</td>\n",
       "      <td>7453</td>\n",
       "      <td>2</td>\n",
       "      <td>49</td>\n",
       "      <td>52</td>\n",
       "      <td>0.013632</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>0.006485</td>\n",
       "      <td>0.006882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>tRNA_435:58</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_tRNA-Val-TAC-2-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>37140</td>\n",
       "      <td>37084</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>7.930725</td>\n",
       "      <td>6313</td>\n",
       "      <td>6193</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>52</td>\n",
       "      <td>0.019008</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.010138</td>\n",
       "      <td>0.008237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>tRNA_436:26</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_tRNA-Val-TAC-3-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>6512</td>\n",
       "      <td>6368</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2.354088</td>\n",
       "      <td>1242</td>\n",
       "      <td>1183</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>0.047504</td>\n",
       "      <td>0.023349</td>\n",
       "      <td>0.000805</td>\n",
       "      <td>0.023349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>tRNA_436:58</td>\n",
       "      <td>.</td>\n",
       "      <td>Homo_sapiens_tRNA-Val-TAC-3-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>2009</td>\n",
       "      <td>2000</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>15.504202</td>\n",
       "      <td>476</td>\n",
       "      <td>431</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>0.094538</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>0.075630</td>\n",
       "      <td>0.016807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>snoRNA_130:86</td>\n",
       "      <td>.</td>\n",
       "      <td>SNORD22;snoID_0137</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>T</td>\n",
       "      <td>14270</td>\n",
       "      <td>14245</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>8.584855</td>\n",
       "      <td>931</td>\n",
       "      <td>916</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0.016112</td>\n",
       "      <td>0.006445</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.001074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Coordinate Comments  \\\n",
       "0     NR_145822.1:12        .   \n",
       "1   NR_145822.1:1322        .   \n",
       "2         tRNA_11:60        .   \n",
       "3         tRNA_41:36        .   \n",
       "4         tRNA_48:36        .   \n",
       "..               ...      ...   \n",
       "86       tRNA_434:58        .   \n",
       "87       tRNA_435:58        .   \n",
       "88       tRNA_436:26        .   \n",
       "89       tRNA_436:58        .   \n",
       "90     snoRNA_130:86        .   \n",
       "\n",
       "                                                 Gene Biotype Strand Base  \\\n",
       "0                                                 28S    rRNA      +    A   \n",
       "1                                                 28S    rRNA      +    A   \n",
       "2   Homo_sapiens_nmt-tRNA-Leu-TAA-1-1 (tRNAscan-SE...    tRNA      +    A   \n",
       "3   Homo_sapiens_tRNA-Ala-AGC-4-1 (tRNAscan-SE ID:...    tRNA      +    A   \n",
       "4   Homo_sapiens_tRNA-Ala-CGC-1-1 (tRNAscan-SE ID:...    tRNA      +    A   \n",
       "..                                                ...     ...    ...  ...   \n",
       "86  Homo_sapiens_tRNA-Val-TAC-1-1 (tRNAscan-SE ID:...    tRNA      +    A   \n",
       "87  Homo_sapiens_tRNA-Val-TAC-2-1 (tRNAscan-SE ID:...    tRNA      +    A   \n",
       "88  Homo_sapiens_tRNA-Val-TAC-3-1 (tRNAscan-SE ID:...    tRNA      +    A   \n",
       "89  Homo_sapiens_tRNA-Val-TAC-3-1 (tRNAscan-SE ID:...    tRNA      +    A   \n",
       "90                                 SNORD22;snoID_0137  snoRNA      +    T   \n",
       "\n",
       "    BS_Depth    BS_A  BS_C  BS_G  ...  Fold (NBS/BS).1  NBS_Depth    NBS_A  \\\n",
       "0     816923  813695   733  1487  ...         2.990034    1023372  1016782   \n",
       "1     197730  188581   102  2959  ...         8.150468      26665    14725   \n",
       "2        387     386     1     0  ...       201.656000        625      170   \n",
       "3       1155    1139     0    13  ...        13.694228        641      489   \n",
       "4        481     462     0    18  ...        50.517572        313      179   \n",
       "..       ...     ...   ...   ...  ...              ...        ...      ...   \n",
       "86     37244   37182     1    37  ...         5.416748       7556     7453   \n",
       "87     37140   37084     3    32  ...         7.930725       6313     6193   \n",
       "88      6512    6368    24     3  ...         2.354088       1242     1183   \n",
       "89      2009    2000     0     7  ...        15.504202        476      431   \n",
       "90     14270   14245     5    10  ...         8.584855        931      916   \n",
       "\n",
       "    NBS_C  NBS_G  NBS_T   NBS_Mut    NBS_%C    NBS_%G    NBS_%T  \n",
       "0    1992   3379   1219  0.006439  0.001947  0.003302  0.001191  \n",
       "1      88   4280   7572  0.447778  0.003300  0.160510  0.283968  \n",
       "2       0     12    443  0.728000  0.000000  0.019200  0.708800  \n",
       "3       8     51     93  0.237129  0.012480  0.079563  0.145086  \n",
       "4       3     39     92  0.428115  0.009585  0.124601  0.293930  \n",
       "..    ...    ...    ...       ...       ...       ...       ...  \n",
       "86      2     49     52  0.013632  0.000265  0.006485  0.006882  \n",
       "87      4     64     52  0.019008  0.000634  0.010138  0.008237  \n",
       "88     29      1     29  0.047504  0.023349  0.000805  0.023349  \n",
       "89      1     36      8  0.094538  0.002101  0.075630  0.016807  \n",
       "90      6      8      1  0.016112  0.006445  0.008593  0.001074  \n",
       "\n",
       "[91 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# not sure how to handle this. Also most of it is tRNA which Sci-Modom can't handle, yet\n",
    "df = pd.read_excel(\"/home/annebusch/anne02/euf-data/rbs-seq/GSE90963_Table_S3-m1A_candidate_sites.xlsx\", sheet_name=\"1. Candidate m1A sites\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cdfe15ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['28S']\n",
      "['28S']\n",
      "['Homo_sapiens_tRNA-Arg-ACG-1-1 (tRNAscan-SE ID', ' chr6.trna6) chr6', '26328368-26328440 (+) Arg (ACG) 73 bp Sc', ' 65.8;Homo_sapiens_tRNA-Arg-ACG-1-2 (tRNAscan-SE ID', ' chr6.trna8) chr6', '26537726-26537798 (+) Arg (ACG) 73 bp Sc', ' 65.8;Homo_sapiens_tRNA-Arg-ACG-1-3 (tRNAscan-SE ID', ' chr14.trna7) chr14', '23398910-23398982 (+) Arg (ACG) 73 bp Sc', ' 65.8']\n",
      "['Homo_sapiens_tRNA-Arg-CCG-1-1 (tRNAscan-SE ID', ' chr6.trna115) chr6', '28710729-28710801 (-) Arg (CCG) 73 bp Sc', ' 64.7;Homo_sapiens_tRNA-Arg-CCG-1-2 (tRNAscan-SE ID', ' chr6.trna72) chr6', '28849165-28849237 (+) Arg (CCG) 73 bp Sc', ' 64.7;Homo_sapiens_tRNA-Arg-CCG-1-3 (tRNAscan-SE ID', ' chr16.trna2) chr16', '3200675-3200747 (+) Arg (CCG) 73 bp Sc', ' 64.7']\n",
      "['Homo_sapiens_tRNA-Asn-GTT-2-1 (tRNAscan-SE ID', ' chr1.trna27) chr1', '148248115-148248188 (+) Asn (GTT) 74 bp Sc', ' 76.2;Homo_sapiens_tRNA-Asn-GTT-2-2 (tRNAscan-SE ID', ' chr1.trna76) chr1', '161397867-161397940 (-) Asn (GTT) 74 bp Sc', ' 76.2;Homo_sapiens_tRNA-Asn-GTT-2-3 (tRNAscan-SE ID', ' chr10.trna5) chr10', '22518438-22518511 (-) Asn (GTT) 74 bp Sc', ' 76.2;Homo_sapiens_tRNA-Asn-GTT-2-4 (tRNAscan-SE ID', ' chr13.trna5) chr13', '31248101-31248174 (-) Asn (GTT) 74 bp Sc', ' 76.2;Homo_sapiens_tRNA-Asn-GTT-2-5 (tRNAscan-SE ID', ' chr17.trna30) chr17', '36908034-36908107 (-) Asn (GTT) 74 bp Sc', ' 76.2;Homo_sapiens_tRNA-Asn-GTT-2-6 (tRNAscan-SE ID', ' chr19.trna1) chr19', '1383562-1383635 (+) Asn (GTT) 74 bp Sc', ' 76.2']\n",
      "['Homo_sapiens_tRNA-Asn-GTT-3-1 (tRNAscan-SE ID', ' chr1.trna98) chr1', '148598314-148598387 (-) Asn (GTT) 74 bp Sc', ' 74.1;Homo_sapiens_tRNA-Asn-GTT-3-2 (tRNAscan-SE ID', ' chr1.trna97) chr1', '148760356-148760429 (-) Asn (GTT) 74 bp Sc', ' 74.1']\n",
      "['Homo_sapiens_tRNA-Gln-CTG-1-1 (tRNAscan-SE ID', ' chr6.trna1) chr6', '18836402-18836473 (+) Gln (CTG) 72 bp Sc', ' 70.6;Homo_sapiens_tRNA-Gln-CTG-1-2 (tRNAscan-SE ID', ' chr6.trna49) chr6', '27487308-27487379 (+) Gln (CTG) 72 bp Sc', ' 70.6;Homo_sapiens_tRNA-Gln-CTG-1-3 (tRNAscan-SE ID', ' chr6.trna99) chr6', '28909378-28909449 (-) Gln (CTG) 72 bp Sc', ' 70.6;Homo_sapiens_tRNA-Gln-CTG-1-4 (tRNAscan-SE ID', ' chr15.trna8) chr15', '66161400-66161471 (-) Gln (CTG) 72 bp Sc', ' 70.6;Homo_sapiens_tRNA-Gln-CTG-1-5 (tRNAscan-SE ID', ' chr17.trna2) chr17', '8023070-8023141 (+) Gln (CTG) 72 bp Sc', ' 70.6']\n",
      "['Homo_sapiens_tRNA-Gly-CCC-1-1 (tRNAscan-SE ID', ' chr1.trna122) chr1', '16872434-16872504 (-) Gly (CCC) 71 bp Sc', ' 75.5;Homo_sapiens_tRNA-Gly-CCC-1-2 (tRNAscan-SE ID', ' chr1.trna4) chr1', '17188416-17188486 (+) Gly (CCC) 71 bp Sc', ' 75.5']\n",
      "['Homo_sapiens_tRNA-Gly-GCC-1-1 (tRNAscan-SE ID', ' chr1.trna34) chr1', '161413094-161413164 (+) Gly (GCC) 71 bp Sc', ' 79.9;Homo_sapiens_tRNA-Gly-GCC-1-2 (tRNAscan-SE ID', ' chr1.trna36) chr1', '161420467-161420537 (+) Gly (GCC) 71 bp Sc', ' 79.9;Homo_sapiens_tRNA-Gly-GCC-1-3 (tRNAscan-SE ID', ' chr1.trna38) chr1', '161427898-161427968 (+) Gly (GCC) 71 bp Sc', ' 79.9;Homo_sapiens_tRNA-Gly-GCC-1-4 (tRNAscan-SE ID', ' chr1.trna40) chr1', '161435258-161435328 (+) Gly (GCC) 71 bp Sc', ' 79.9;Homo_sapiens_tRNA-Gly-GCC-1-5 (tRNAscan-SE ID', ' chr21.trna1) chr21', '18827107-18827177 (-) Gly (GCC) 71 bp Sc', ' 79.9']\n",
      "['Homo_sapiens_tRNA-Gly-GCC-2-1 (tRNAscan-SE ID', ' chr1.trna61) chr1', '161493637-161493707 (-) Gly (GCC) 71 bp Sc', ' 77.9;Homo_sapiens_tRNA-Gly-GCC-2-2 (tRNAscan-SE ID', ' chr2.trna17) chr2', '157257659-157257729 (-) Gly (GCC) 71 bp Sc', ' 77.9;Homo_sapiens_tRNA-Gly-GCC-2-3 (tRNAscan-SE ID', ' chr6.trna129) chr6', '27870686-27870756 (-) Gly (GCC) 71 bp Sc', ' 77.9;Homo_sapiens_tRNA-Gly-GCC-2-4 (tRNAscan-SE ID', ' chr16.trna24) chr16', '70812942-70813012 (-) Gly (GCC) 71 bp Sc', ' 77.9;Homo_sapiens_tRNA-Gly-GCC-2-5 (tRNAscan-SE ID', ' chr16.trna20) chr16', '70823410-70823480 (+) Gly (GCC) 71 bp Sc', ' 77.9;Homo_sapiens_tRNA-Gly-GCC-2-6 (tRNAscan-SE ID', ' chr17.trna4) chr17', '8029064-8029134 (+) Gly (GCC) 71 bp Sc', ' 77.9']\n",
      "['Homo_sapiens_tRNA-Gly-TCC-2-1 (tRNAscan-SE ID', ' chr1.trna107) chr1', '145397864-145397935 (-) Gly (TCC) 72 bp Sc', ' 71.6;Homo_sapiens_tRNA-Gly-TCC-2-2 (tRNAscan-SE ID', ' chr1.trna72) chr1', '161417375-161417446 (-) Gly (TCC) 72 bp Sc', ' 71.6;Homo_sapiens_tRNA-Gly-TCC-2-3 (tRNAscan-SE ID', ' chr1.trna69) chr1', '161424756-161424827 (-) Gly (TCC) 72 bp Sc', ' 71.6;Homo_sapiens_tRNA-Gly-TCC-2-4 (tRNAscan-SE ID', ' chr1.trna66) chr1', '161432166-161432237 (-) Gly (TCC) 72 bp Sc', ' 71.6;Homo_sapiens_tRNA-Gly-TCC-2-5 (tRNAscan-SE ID', ' chr1.trna63) chr1', '161439547-161439618 (-) Gly (TCC) 72 bp Sc', ' 71.6;Homo_sapiens_tRNA-Gly-TCC-2-6 (tRNAscan-SE ID', ' chr1.trna43) chr1', '161500903-161500974 (+) Gly (TCC) 72 bp Sc', ' 71.6']\n",
      "['Homo_sapiens_tRNA-His-GTG-1-1 (tRNAscan-SE ID', ' chr1.trna108) chr1', '145396881-145396952 (-) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-2 (tRNAscan-SE ID', ' chr1.trna18) chr1', '146544773-146544844 (+) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-3 (tRNAscan-SE ID', ' chr1.trna22) chr1', '147753471-147753542 (+) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-4 (tRNAscan-SE ID', ' chr1.trna100) chr1', '147774845-147774916 (-) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-5 (tRNAscan-SE ID', ' chr6.trna32) chr6', '27125906-27125977 (+) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-6 (tRNAscan-SE ID', ' chr9.trna9) chr9', '14433938-14434009 (-) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-7 (tRNAscan-SE ID', ' chr15.trna10) chr15', '45490804-45490875 (-) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-8 (tRNAscan-SE ID', ' chr15.trna9) chr15', '45492611-45492682 (-) His (GTG) 72 bp Sc', ' 60.6;Homo_sapiens_tRNA-His-GTG-1-9 (tRNAscan-SE ID', ' chr15.trna1) chr15', '45493349-45493420 (+) His (GTG) 72 bp Sc', ' 60.6']\n",
      "['Homo_sapiens_tRNA-Ile-AAT-5-1 (tRNAscan-SE ID', ' chr6.trna11) chr6', '26554350-26554423 (+) Ile (AAT) 74 bp Sc', ' 80.0;Homo_sapiens_tRNA-Ile-AAT-5-2 (tRNAscan-SE ID', ' chr6.trna159) chr6', '27144994-27145067 (-) Ile (AAT) 74 bp Sc', ' 80.0;Homo_sapiens_tRNA-Ile-AAT-5-3 (tRNAscan-SE ID', ' chr6.trna155) chr6', '27205350-27205423 (-) Ile (AAT) 74 bp Sc', ' 80.0;Homo_sapiens_tRNA-Ile-AAT-5-4 (tRNAscan-SE ID', ' chr14.trna12) chr14', '102783429-102783502 (+) Ile (AAT) 74 bp Sc', ' 80.0;Homo_sapiens_tRNA-Ile-AAT-5-5 (tRNAscan-SE ID', ' chr17.trna8) chr17', '8090911-8090984 (+) Ile (AAT) 74 bp Sc', ' 80.0']\n",
      "['Homo_sapiens_tRNA-Ile-AAT-7-1 (tRNAscan-SE ID', ' chr6.trna166) chr6', '26721221-26721294 (-) Ile (AAT) 74 bp Sc', ' 80.0;Homo_sapiens_tRNA-Ile-AAT-7-2 (tRNAscan-SE ID', ' chr6.trna27) chr6', '26780832-26780905 (+) Ile (AAT) 74 bp Sc', ' 80.0']\n",
      "['Homo_sapiens_tRNA-iMet-CAT-1-1 (tRNAscan-SE ID', ' chr1.trna31) chr1', '153643726-153643797 (+) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-2 (tRNAscan-SE ID', ' chr6.trna2) chr6', '26286754-26286825 (+) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-3 (tRNAscan-SE ID', ' chr6.trna172) chr6', '26313352-26313423 (-) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-4 (tRNAscan-SE ID', ' chr6.trna170) chr6', '26330529-26330600 (-) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-5 (tRNAscan-SE ID', ' chr6.trna151) chr6', '27300764-27300835 (-) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-6 (tRNAscan-SE ID', ' chr6.trna143) chr6', '27560600-27560671 (-) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-7 (tRNAscan-SE ID', ' chr6.trna130) chr6', '27870271-27870342 (-) Met (CAT) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-iMet-CAT-1-8 (tRNAscan-SE ID', ' chr17.trna18) chr17', '80452597-80452668 (-) Met (CAT) 72 bp Sc', ' 67.7']\n",
      "['Homo_sapiens_tRNA-Leu-AAG-1-1 (tRNAscan-SE ID', ' chr5.trna22) chr5', '180524474-180524555 (-) Leu (AAG) 82 bp Sc', ' 76.4;Homo_sapiens_tRNA-Leu-AAG-1-2 (tRNAscan-SE ID', ' chr5.trna6) chr5', '180528840-180528921 (+) Leu (AAG) 82 bp Sc', ' 76.4;Homo_sapiens_tRNA-Leu-AAG-1-3 (tRNAscan-SE ID', ' chr5.trna19) chr5', '180601044-180601125 (-) Leu (AAG) 82 bp Sc', ' 76.4']\n",
      "['Homo_sapiens_tRNA-Leu-CAG-1-1 (tRNAscan-SE ID', ' chr1.trna33) chr1', '161411323-161411405 (+) Leu (CAG) 83 bp Sc', ' 83.1;Homo_sapiens_tRNA-Leu-CAG-1-2 (tRNAscan-SE ID', ' chr1.trna35) chr1', '161418741-161418823 (+) Leu (CAG) 83 bp Sc', ' 83.1;Homo_sapiens_tRNA-Leu-CAG-1-3 (tRNAscan-SE ID', ' chr1.trna37) chr1', '161426122-161426204 (+) Leu (CAG) 83 bp Sc', ' 83.1;Homo_sapiens_tRNA-Leu-CAG-1-4 (tRNAscan-SE ID', ' chr1.trna39) chr1', '161433532-161433614 (+) Leu (CAG) 83 bp Sc', ' 83.1;Homo_sapiens_tRNA-Leu-CAG-1-5 (tRNAscan-SE ID', ' chr1.trna41) chr1', '161440913-161440995 (+) Leu (CAG) 83 bp Sc', ' 83.1;Homo_sapiens_tRNA-Leu-CAG-1-6 (tRNAscan-SE ID', ' chr1.trna60) chr1', '161500132-161500214 (-) Leu (CAG) 83 bp Sc', ' 83.1;Homo_sapiens_tRNA-Leu-CAG-1-7 (tRNAscan-SE ID', ' chr6.trna7) chr6', '26521436-26521518 (+) Leu (CAG) 83 bp Sc', ' 83.1']\n",
      "['Homo_sapiens_tRNA-Leu-CAG-2-1 (tRNAscan-SE ID', ' chr16.trna18) chr16', '57333863-57333945 (+) Leu (CAG) 83 bp Sc', ' 82.4;Homo_sapiens_tRNA-Leu-CAG-2-2 (tRNAscan-SE ID', ' chr16.trna26) chr16', '57334392-57334474 (-) Leu (CAG) 83 bp Sc', ' 82.4']\n",
      "['Homo_sapiens_tRNA-Lys-CTT-1-1 (tRNAscan-SE ID', ' chr14.trna14) chr14', '58706613-58706685 (-) Lys (CTT) 73 bp Sc', ' 78.5;Homo_sapiens_tRNA-Lys-CTT-1-2 (tRNAscan-SE ID', ' chr15.trna2) chr15', '79152904-79152976 (+) Lys (CTT) 73 bp Sc', ' 78.5']\n",
      "['Homo_sapiens_tRNA-Lys-CTT-2-1 (tRNAscan-SE ID', ' chr1.trna109) chr1', '145395522-145395594 (-) Lys (CTT) 73 bp Sc', ' 78.0;Homo_sapiens_tRNA-Lys-CTT-2-2 (tRNAscan-SE ID', ' chr5.trna12) chr5', '180634755-180634827 (+) Lys (CTT) 73 bp Sc', ' 78.0;Homo_sapiens_tRNA-Lys-CTT-2-3 (tRNAscan-SE ID', ' chr5.trna14) chr5', '180648979-180649051 (-) Lys (CTT) 73 bp Sc', ' 78.0;Homo_sapiens_tRNA-Lys-CTT-2-4 (tRNAscan-SE ID', ' chr6.trna13) chr6', '26556774-26556846 (+) Lys (CTT) 73 bp Sc', ' 78.0;Homo_sapiens_tRNA-Lys-CTT-2-5 (tRNAscan-SE ID', ' chr16.trna8) chr16', '3225692-3225764 (+) Lys (CTT) 73 bp Sc', ' 78.0']\n",
      "['Homo_sapiens_tRNA-Met-CAT-3-1 (tRNAscan-SE ID', ' chr6.trna74) chr6', '28912352-28912424 (+) Met (CAT) 73 bp Sc', ' 74.0;Homo_sapiens_tRNA-Met-CAT-3-2 (tRNAscan-SE ID', ' chr6.trna97) chr6', '28921042-28921114 (-) Met (CAT) 73 bp Sc', ' 74.0']\n",
      "['Homo_sapiens_tRNA-Phe-GAA-1-1 (tRNAscan-SE ID', ' chr6.trna110) chr6', '28758499-28758571 (-) Phe (GAA) 73 bp Sc', ' 85.2;Homo_sapiens_tRNA-Phe-GAA-1-2 (tRNAscan-SE ID', ' chr6.trna96) chr6', '28949449-28949521 (-) Phe (GAA) 73 bp Sc', ' 85.2;Homo_sapiens_tRNA-Phe-GAA-1-3 (tRNAscan-SE ID', ' chr11.trna17) chr11', '59324970-59325042 (-) Phe (GAA) 73 bp Sc', ' 85.2;Homo_sapiens_tRNA-Phe-GAA-1-4 (tRNAscan-SE ID', ' chr12.trna11) chr12', '125412389-125412461 (-) Phe (GAA) 73 bp Sc', ' 85.2;Homo_sapiens_tRNA-Phe-GAA-1-5 (tRNAscan-SE ID', ' chr13.trna1) chr13', '95201904-95201976 (-) Phe (GAA) 73 bp Sc', ' 85.2;Homo_sapiens_tRNA-Phe-GAA-1-6 (tRNAscan-SE ID', ' chr19.trna14) chr19', '1383361-1383433 (-) Phe (GAA) 73 bp Sc', ' 85.2']\n",
      "['Homo_sapiens_tRNA-Pro-AGG-2-1 (tRNAscan-SE ID', ' chr1.trna58) chr1', '167684725-167684796 (-) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-2 (tRNAscan-SE ID', ' chr6.trna12) chr6', '26555498-26555569 (+) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-3 (tRNAscan-SE ID', ' chr7.trna4) chr7', '128423504-128423575 (+) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-4 (tRNAscan-SE ID', ' chr11.trna7) chr11', '75946557-75946628 (+) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-5 (tRNAscan-SE ID', ' chr14.trna24) chr14', '21077495-21077566 (-) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-6 (tRNAscan-SE ID', ' chr14.trna23) chr14', '21081560-21081631 (-) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-7 (tRNAscan-SE ID', ' chr16.trna29) chr16', '3232635-3232706 (-) Pro (AGG) 72 bp Sc', ' 67.7;Homo_sapiens_tRNA-Pro-AGG-2-8 (tRNAscan-SE ID', ' chr16.trna10) chr16', '3239634-3239705 (+) Pro (AGG) 72 bp Sc', ' 67.7']\n",
      "['Homo_sapiens_tRNA-Ser-AGA-2-1 (tRNAscan-SE ID', ' chr6.trna5) chr6', '26327817-26327898 (+) Ser (AGA) 82 bp Sc', ' 89.6;Homo_sapiens_tRNA-Ser-AGA-2-2 (tRNAscan-SE ID', ' chr6.trna44) chr6', '27446591-27446672 (+) Ser (AGA) 82 bp Sc', ' 89.6;Homo_sapiens_tRNA-Ser-AGA-2-3 (tRNAscan-SE ID', ' chr6.trna46) chr6', '27463593-27463674 (+) Ser (AGA) 82 bp Sc', ' 89.6;Homo_sapiens_tRNA-Ser-AGA-2-4 (tRNAscan-SE ID', ' chr6.trna47) chr6', '27470818-27470899 (+) Ser (AGA) 82 bp Sc', ' 89.6;Homo_sapiens_tRNA-Ser-AGA-2-5 (tRNAscan-SE ID', ' chr8.trna9) chr8', '96281885-96281966 (-) Ser (AGA) 82 bp Sc', ' 89.6;Homo_sapiens_tRNA-Ser-AGA-2-6 (tRNAscan-SE ID', ' chr17.trna35) chr17', '8129928-8130009 (-) Ser (AGA) 82 bp Sc', ' 89.6']\n",
      "['Homo_sapiens_tRNA-Ser-GCT-4-1 (tRNAscan-SE ID', ' chr6.trna124) chr6', '28565117-28565198 (-) Ser (GCT) 82 bp Sc', ' 88.1;Homo_sapiens_tRNA-Ser-GCT-4-2 (tRNAscan-SE ID', ' chr15.trna11) chr15', '40886023-40886104 (-) Ser (GCT) 82 bp Sc', ' 88.1;Homo_sapiens_tRNA-Ser-GCT-4-3 (tRNAscan-SE ID', ' chr17.trna6) chr17', '8090184-8090265 (+) Ser (GCT) 82 bp Sc', ' 88.1']\n",
      "['Homo_sapiens_tRNA-Val-CAC-1-1 (tRNAscan-SE ID', ' chr1.trna78) chr1', '161369490-161369562 (-) Val (CAC) 73 bp Sc', ' 78.8;Homo_sapiens_tRNA-Val-CAC-1-2 (tRNAscan-SE ID', ' chr5.trna5) chr5', '180524070-180524142 (+) Val (CAC) 73 bp Sc', ' 78.8;Homo_sapiens_tRNA-Val-CAC-1-3 (tRNAscan-SE ID', ' chr5.trna21) chr5', '180529253-180529325 (-) Val (CAC) 73 bp Sc', ' 78.8;Homo_sapiens_tRNA-Val-CAC-1-4 (tRNAscan-SE ID', ' chr5.trna9) chr5', '180600650-180600722 (+) Val (CAC) 73 bp Sc', ' 78.8;Homo_sapiens_tRNA-Val-CAC-1-5 (tRNAscan-SE ID', ' chr5.trna13) chr5', '180649395-180649467 (-) Val (CAC) 73 bp Sc', ' 78.8;Homo_sapiens_tRNA-Val-CAC-1-6 (tRNAscan-SE ID', ' chr6.trna9) chr6', '26538282-26538354 (+) Val (CAC) 73 bp Sc', ' 78.8']\n",
      "['Homo_sapiens_tRNA-Val-TAC-1-1 (tRNAscan-SE ID', ' chr11.trna19) chr11', '59318102-59318174 (-) Val (TAC) 73 bp Sc', ' 78.8;Homo_sapiens_tRNA-Val-TAC-1-2 (tRNAscan-SE ID', ' chrX.trna4) chrX', '18693029-18693101 (-) Val (TAC) 73 bp Sc', ' 78.8']\n",
      "['SNORD22;snoID_0137']\n"
     ]
    }
   ],
   "source": [
    "for row in df[\"Gene\"]:\n",
    "    \n",
    "    if len(row.split(\":\")) == 4:\n",
    "        # print(row.split(\":\"))\n",
    "        homo, chrom, positions, sc = row.split(\":\")\n",
    "        \n",
    "    else: # if longer than 4, then there are two exons\n",
    "        print(row.split(\":\"))\n",
    "    row = row.split(\":\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c5d4897",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens_tRNA-Arg-ACG-1-1 (tRNAscan-SE ID: chr6.trna6) chr6:26328368-26328440 (+) Arg (ACG) 73 bp Sc: 65.8;Homo_sapiens_tRNA-Arg-ACG-1-2 (tRNAscan-SE ID: chr6.trna8) chr6:26537726-26537798 (+) Arg (ACG) 73 bp Sc: 65.8;Homo_sapiens_tRNA-Arg-ACG-1-3 (tRNAscan-SE ID: chr14.trna7) chr14:23398910-23398982 (+) Arg (ACG) 73 bp Sc: 65.8\n",
      "-\n",
      "Homo_sapiens_tRNA-Arg-CCG-1-1 (tRNAscan-SE ID: chr6.trna115) chr6:28710729-28710801 (-) Arg (CCG) 73 bp Sc: 64.7;Homo_sapiens_tRNA-Arg-CCG-1-2 (tRNAscan-SE ID: chr6.trna72) chr6:28849165-28849237 (+) Arg (CCG) 73 bp Sc: 64.7;Homo_sapiens_tRNA-Arg-CCG-1-3 (tRNAscan-SE ID: chr16.trna2) chr16:3200675-3200747 (+) Arg (CCG) 73 bp Sc: 64.7\n",
      "-\n",
      "Homo_sapiens_tRNA-Asn-GTT-2-1 (tRNAscan-SE ID: chr1.trna27) chr1:148248115-148248188 (+) Asn (GTT) 74 bp Sc: 76.2;Homo_sapiens_tRNA-Asn-GTT-2-2 (tRNAscan-SE ID: chr1.trna76) chr1:161397867-161397940 (-) Asn (GTT) 74 bp Sc: 76.2;Homo_sapiens_tRNA-Asn-GTT-2-3 (tRNAscan-SE ID: chr10.trna5) chr10:22518438-22518511 (-) Asn (GTT) 74 bp Sc: 76.2;Homo_sapiens_tRNA-Asn-GTT-2-4 (tRNAscan-SE ID: chr13.trna5) chr13:31248101-31248174 (-) Asn (GTT) 74 bp Sc: 76.2;Homo_sapiens_tRNA-Asn-GTT-2-5 (tRNAscan-SE ID: chr17.trna30) chr17:36908034-36908107 (-) Asn (GTT) 74 bp Sc: 76.2;Homo_sapiens_tRNA-Asn-GTT-2-6 (tRNAscan-SE ID: chr19.trna1) chr19:1383562-1383635 (+) Asn (GTT) 74 bp Sc: 76.2\n",
      "-\n",
      "Homo_sapiens_tRNA-Asn-GTT-3-1 (tRNAscan-SE ID: chr1.trna98) chr1:148598314-148598387 (-) Asn (GTT) 74 bp Sc: 74.1;Homo_sapiens_tRNA-Asn-GTT-3-2 (tRNAscan-SE ID: chr1.trna97) chr1:148760356-148760429 (-) Asn (GTT) 74 bp Sc: 74.1\n",
      "-\n",
      "Homo_sapiens_tRNA-Gln-CTG-1-1 (tRNAscan-SE ID: chr6.trna1) chr6:18836402-18836473 (+) Gln (CTG) 72 bp Sc: 70.6;Homo_sapiens_tRNA-Gln-CTG-1-2 (tRNAscan-SE ID: chr6.trna49) chr6:27487308-27487379 (+) Gln (CTG) 72 bp Sc: 70.6;Homo_sapiens_tRNA-Gln-CTG-1-3 (tRNAscan-SE ID: chr6.trna99) chr6:28909378-28909449 (-) Gln (CTG) 72 bp Sc: 70.6;Homo_sapiens_tRNA-Gln-CTG-1-4 (tRNAscan-SE ID: chr15.trna8) chr15:66161400-66161471 (-) Gln (CTG) 72 bp Sc: 70.6;Homo_sapiens_tRNA-Gln-CTG-1-5 (tRNAscan-SE ID: chr17.trna2) chr17:8023070-8023141 (+) Gln (CTG) 72 bp Sc: 70.6\n",
      "-\n",
      "Homo_sapiens_tRNA-Gly-CCC-1-1 (tRNAscan-SE ID: chr1.trna122) chr1:16872434-16872504 (-) Gly (CCC) 71 bp Sc: 75.5;Homo_sapiens_tRNA-Gly-CCC-1-2 (tRNAscan-SE ID: chr1.trna4) chr1:17188416-17188486 (+) Gly (CCC) 71 bp Sc: 75.5\n",
      "-\n",
      "Homo_sapiens_tRNA-Gly-GCC-1-1 (tRNAscan-SE ID: chr1.trna34) chr1:161413094-161413164 (+) Gly (GCC) 71 bp Sc: 79.9;Homo_sapiens_tRNA-Gly-GCC-1-2 (tRNAscan-SE ID: chr1.trna36) chr1:161420467-161420537 (+) Gly (GCC) 71 bp Sc: 79.9;Homo_sapiens_tRNA-Gly-GCC-1-3 (tRNAscan-SE ID: chr1.trna38) chr1:161427898-161427968 (+) Gly (GCC) 71 bp Sc: 79.9;Homo_sapiens_tRNA-Gly-GCC-1-4 (tRNAscan-SE ID: chr1.trna40) chr1:161435258-161435328 (+) Gly (GCC) 71 bp Sc: 79.9;Homo_sapiens_tRNA-Gly-GCC-1-5 (tRNAscan-SE ID: chr21.trna1) chr21:18827107-18827177 (-) Gly (GCC) 71 bp Sc: 79.9\n",
      "-\n",
      "Homo_sapiens_tRNA-Gly-GCC-2-1 (tRNAscan-SE ID: chr1.trna61) chr1:161493637-161493707 (-) Gly (GCC) 71 bp Sc: 77.9;Homo_sapiens_tRNA-Gly-GCC-2-2 (tRNAscan-SE ID: chr2.trna17) chr2:157257659-157257729 (-) Gly (GCC) 71 bp Sc: 77.9;Homo_sapiens_tRNA-Gly-GCC-2-3 (tRNAscan-SE ID: chr6.trna129) chr6:27870686-27870756 (-) Gly (GCC) 71 bp Sc: 77.9;Homo_sapiens_tRNA-Gly-GCC-2-4 (tRNAscan-SE ID: chr16.trna24) chr16:70812942-70813012 (-) Gly (GCC) 71 bp Sc: 77.9;Homo_sapiens_tRNA-Gly-GCC-2-5 (tRNAscan-SE ID: chr16.trna20) chr16:70823410-70823480 (+) Gly (GCC) 71 bp Sc: 77.9;Homo_sapiens_tRNA-Gly-GCC-2-6 (tRNAscan-SE ID: chr17.trna4) chr17:8029064-8029134 (+) Gly (GCC) 71 bp Sc: 77.9\n",
      "-\n",
      "Homo_sapiens_tRNA-Gly-TCC-2-1 (tRNAscan-SE ID: chr1.trna107) chr1:145397864-145397935 (-) Gly (TCC) 72 bp Sc: 71.6;Homo_sapiens_tRNA-Gly-TCC-2-2 (tRNAscan-SE ID: chr1.trna72) chr1:161417375-161417446 (-) Gly (TCC) 72 bp Sc: 71.6;Homo_sapiens_tRNA-Gly-TCC-2-3 (tRNAscan-SE ID: chr1.trna69) chr1:161424756-161424827 (-) Gly (TCC) 72 bp Sc: 71.6;Homo_sapiens_tRNA-Gly-TCC-2-4 (tRNAscan-SE ID: chr1.trna66) chr1:161432166-161432237 (-) Gly (TCC) 72 bp Sc: 71.6;Homo_sapiens_tRNA-Gly-TCC-2-5 (tRNAscan-SE ID: chr1.trna63) chr1:161439547-161439618 (-) Gly (TCC) 72 bp Sc: 71.6;Homo_sapiens_tRNA-Gly-TCC-2-6 (tRNAscan-SE ID: chr1.trna43) chr1:161500903-161500974 (+) Gly (TCC) 72 bp Sc: 71.6\n",
      "-\n",
      "Homo_sapiens_tRNA-His-GTG-1-1 (tRNAscan-SE ID: chr1.trna108) chr1:145396881-145396952 (-) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-2 (tRNAscan-SE ID: chr1.trna18) chr1:146544773-146544844 (+) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-3 (tRNAscan-SE ID: chr1.trna22) chr1:147753471-147753542 (+) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-4 (tRNAscan-SE ID: chr1.trna100) chr1:147774845-147774916 (-) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-5 (tRNAscan-SE ID: chr6.trna32) chr6:27125906-27125977 (+) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-6 (tRNAscan-SE ID: chr9.trna9) chr9:14433938-14434009 (-) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-7 (tRNAscan-SE ID: chr15.trna10) chr15:45490804-45490875 (-) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-8 (tRNAscan-SE ID: chr15.trna9) chr15:45492611-45492682 (-) His (GTG) 72 bp Sc: 60.6;Homo_sapiens_tRNA-His-GTG-1-9 (tRNAscan-SE ID: chr15.trna1) chr15:45493349-45493420 (+) His (GTG) 72 bp Sc: 60.6\n",
      "-\n",
      "Homo_sapiens_tRNA-Ile-AAT-5-1 (tRNAscan-SE ID: chr6.trna11) chr6:26554350-26554423 (+) Ile (AAT) 74 bp Sc: 80.0;Homo_sapiens_tRNA-Ile-AAT-5-2 (tRNAscan-SE ID: chr6.trna159) chr6:27144994-27145067 (-) Ile (AAT) 74 bp Sc: 80.0;Homo_sapiens_tRNA-Ile-AAT-5-3 (tRNAscan-SE ID: chr6.trna155) chr6:27205350-27205423 (-) Ile (AAT) 74 bp Sc: 80.0;Homo_sapiens_tRNA-Ile-AAT-5-4 (tRNAscan-SE ID: chr14.trna12) chr14:102783429-102783502 (+) Ile (AAT) 74 bp Sc: 80.0;Homo_sapiens_tRNA-Ile-AAT-5-5 (tRNAscan-SE ID: chr17.trna8) chr17:8090911-8090984 (+) Ile (AAT) 74 bp Sc: 80.0\n",
      "-\n",
      "Homo_sapiens_tRNA-Ile-AAT-7-1 (tRNAscan-SE ID: chr6.trna166) chr6:26721221-26721294 (-) Ile (AAT) 74 bp Sc: 80.0;Homo_sapiens_tRNA-Ile-AAT-7-2 (tRNAscan-SE ID: chr6.trna27) chr6:26780832-26780905 (+) Ile (AAT) 74 bp Sc: 80.0\n",
      "-\n",
      "Homo_sapiens_tRNA-iMet-CAT-1-1 (tRNAscan-SE ID: chr1.trna31) chr1:153643726-153643797 (+) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-2 (tRNAscan-SE ID: chr6.trna2) chr6:26286754-26286825 (+) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-3 (tRNAscan-SE ID: chr6.trna172) chr6:26313352-26313423 (-) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-4 (tRNAscan-SE ID: chr6.trna170) chr6:26330529-26330600 (-) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-5 (tRNAscan-SE ID: chr6.trna151) chr6:27300764-27300835 (-) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-6 (tRNAscan-SE ID: chr6.trna143) chr6:27560600-27560671 (-) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-7 (tRNAscan-SE ID: chr6.trna130) chr6:27870271-27870342 (-) Met (CAT) 72 bp Sc: 67.7;Homo_sapiens_tRNA-iMet-CAT-1-8 (tRNAscan-SE ID: chr17.trna18) chr17:80452597-80452668 (-) Met (CAT) 72 bp Sc: 67.7\n",
      "-\n",
      "Homo_sapiens_tRNA-Leu-AAG-1-1 (tRNAscan-SE ID: chr5.trna22) chr5:180524474-180524555 (-) Leu (AAG) 82 bp Sc: 76.4;Homo_sapiens_tRNA-Leu-AAG-1-2 (tRNAscan-SE ID: chr5.trna6) chr5:180528840-180528921 (+) Leu (AAG) 82 bp Sc: 76.4;Homo_sapiens_tRNA-Leu-AAG-1-3 (tRNAscan-SE ID: chr5.trna19) chr5:180601044-180601125 (-) Leu (AAG) 82 bp Sc: 76.4\n",
      "-\n",
      "Homo_sapiens_tRNA-Leu-CAG-1-1 (tRNAscan-SE ID: chr1.trna33) chr1:161411323-161411405 (+) Leu (CAG) 83 bp Sc: 83.1;Homo_sapiens_tRNA-Leu-CAG-1-2 (tRNAscan-SE ID: chr1.trna35) chr1:161418741-161418823 (+) Leu (CAG) 83 bp Sc: 83.1;Homo_sapiens_tRNA-Leu-CAG-1-3 (tRNAscan-SE ID: chr1.trna37) chr1:161426122-161426204 (+) Leu (CAG) 83 bp Sc: 83.1;Homo_sapiens_tRNA-Leu-CAG-1-4 (tRNAscan-SE ID: chr1.trna39) chr1:161433532-161433614 (+) Leu (CAG) 83 bp Sc: 83.1;Homo_sapiens_tRNA-Leu-CAG-1-5 (tRNAscan-SE ID: chr1.trna41) chr1:161440913-161440995 (+) Leu (CAG) 83 bp Sc: 83.1;Homo_sapiens_tRNA-Leu-CAG-1-6 (tRNAscan-SE ID: chr1.trna60) chr1:161500132-161500214 (-) Leu (CAG) 83 bp Sc: 83.1;Homo_sapiens_tRNA-Leu-CAG-1-7 (tRNAscan-SE ID: chr6.trna7) chr6:26521436-26521518 (+) Leu (CAG) 83 bp Sc: 83.1\n",
      "-\n",
      "Homo_sapiens_tRNA-Leu-CAG-2-1 (tRNAscan-SE ID: chr16.trna18) chr16:57333863-57333945 (+) Leu (CAG) 83 bp Sc: 82.4;Homo_sapiens_tRNA-Leu-CAG-2-2 (tRNAscan-SE ID: chr16.trna26) chr16:57334392-57334474 (-) Leu (CAG) 83 bp Sc: 82.4\n",
      "-\n",
      "Homo_sapiens_tRNA-Lys-CTT-1-1 (tRNAscan-SE ID: chr14.trna14) chr14:58706613-58706685 (-) Lys (CTT) 73 bp Sc: 78.5;Homo_sapiens_tRNA-Lys-CTT-1-2 (tRNAscan-SE ID: chr15.trna2) chr15:79152904-79152976 (+) Lys (CTT) 73 bp Sc: 78.5\n",
      "-\n",
      "Homo_sapiens_tRNA-Lys-CTT-2-1 (tRNAscan-SE ID: chr1.trna109) chr1:145395522-145395594 (-) Lys (CTT) 73 bp Sc: 78.0;Homo_sapiens_tRNA-Lys-CTT-2-2 (tRNAscan-SE ID: chr5.trna12) chr5:180634755-180634827 (+) Lys (CTT) 73 bp Sc: 78.0;Homo_sapiens_tRNA-Lys-CTT-2-3 (tRNAscan-SE ID: chr5.trna14) chr5:180648979-180649051 (-) Lys (CTT) 73 bp Sc: 78.0;Homo_sapiens_tRNA-Lys-CTT-2-4 (tRNAscan-SE ID: chr6.trna13) chr6:26556774-26556846 (+) Lys (CTT) 73 bp Sc: 78.0;Homo_sapiens_tRNA-Lys-CTT-2-5 (tRNAscan-SE ID: chr16.trna8) chr16:3225692-3225764 (+) Lys (CTT) 73 bp Sc: 78.0\n",
      "-\n",
      "Homo_sapiens_tRNA-Met-CAT-3-1 (tRNAscan-SE ID: chr6.trna74) chr6:28912352-28912424 (+) Met (CAT) 73 bp Sc: 74.0;Homo_sapiens_tRNA-Met-CAT-3-2 (tRNAscan-SE ID: chr6.trna97) chr6:28921042-28921114 (-) Met (CAT) 73 bp Sc: 74.0\n",
      "-\n",
      "Homo_sapiens_tRNA-Phe-GAA-1-1 (tRNAscan-SE ID: chr6.trna110) chr6:28758499-28758571 (-) Phe (GAA) 73 bp Sc: 85.2;Homo_sapiens_tRNA-Phe-GAA-1-2 (tRNAscan-SE ID: chr6.trna96) chr6:28949449-28949521 (-) Phe (GAA) 73 bp Sc: 85.2;Homo_sapiens_tRNA-Phe-GAA-1-3 (tRNAscan-SE ID: chr11.trna17) chr11:59324970-59325042 (-) Phe (GAA) 73 bp Sc: 85.2;Homo_sapiens_tRNA-Phe-GAA-1-4 (tRNAscan-SE ID: chr12.trna11) chr12:125412389-125412461 (-) Phe (GAA) 73 bp Sc: 85.2;Homo_sapiens_tRNA-Phe-GAA-1-5 (tRNAscan-SE ID: chr13.trna1) chr13:95201904-95201976 (-) Phe (GAA) 73 bp Sc: 85.2;Homo_sapiens_tRNA-Phe-GAA-1-6 (tRNAscan-SE ID: chr19.trna14) chr19:1383361-1383433 (-) Phe (GAA) 73 bp Sc: 85.2\n",
      "-\n",
      "Homo_sapiens_tRNA-Pro-AGG-2-1 (tRNAscan-SE ID: chr1.trna58) chr1:167684725-167684796 (-) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-2 (tRNAscan-SE ID: chr6.trna12) chr6:26555498-26555569 (+) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-3 (tRNAscan-SE ID: chr7.trna4) chr7:128423504-128423575 (+) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-4 (tRNAscan-SE ID: chr11.trna7) chr11:75946557-75946628 (+) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-5 (tRNAscan-SE ID: chr14.trna24) chr14:21077495-21077566 (-) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-6 (tRNAscan-SE ID: chr14.trna23) chr14:21081560-21081631 (-) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-7 (tRNAscan-SE ID: chr16.trna29) chr16:3232635-3232706 (-) Pro (AGG) 72 bp Sc: 67.7;Homo_sapiens_tRNA-Pro-AGG-2-8 (tRNAscan-SE ID: chr16.trna10) chr16:3239634-3239705 (+) Pro (AGG) 72 bp Sc: 67.7\n",
      "-\n",
      "Homo_sapiens_tRNA-Ser-AGA-2-1 (tRNAscan-SE ID: chr6.trna5) chr6:26327817-26327898 (+) Ser (AGA) 82 bp Sc: 89.6;Homo_sapiens_tRNA-Ser-AGA-2-2 (tRNAscan-SE ID: chr6.trna44) chr6:27446591-27446672 (+) Ser (AGA) 82 bp Sc: 89.6;Homo_sapiens_tRNA-Ser-AGA-2-3 (tRNAscan-SE ID: chr6.trna46) chr6:27463593-27463674 (+) Ser (AGA) 82 bp Sc: 89.6;Homo_sapiens_tRNA-Ser-AGA-2-4 (tRNAscan-SE ID: chr6.trna47) chr6:27470818-27470899 (+) Ser (AGA) 82 bp Sc: 89.6;Homo_sapiens_tRNA-Ser-AGA-2-5 (tRNAscan-SE ID: chr8.trna9) chr8:96281885-96281966 (-) Ser (AGA) 82 bp Sc: 89.6;Homo_sapiens_tRNA-Ser-AGA-2-6 (tRNAscan-SE ID: chr17.trna35) chr17:8129928-8130009 (-) Ser (AGA) 82 bp Sc: 89.6\n",
      "-\n",
      "Homo_sapiens_tRNA-Ser-GCT-4-1 (tRNAscan-SE ID: chr6.trna124) chr6:28565117-28565198 (-) Ser (GCT) 82 bp Sc: 88.1;Homo_sapiens_tRNA-Ser-GCT-4-2 (tRNAscan-SE ID: chr15.trna11) chr15:40886023-40886104 (-) Ser (GCT) 82 bp Sc: 88.1;Homo_sapiens_tRNA-Ser-GCT-4-3 (tRNAscan-SE ID: chr17.trna6) chr17:8090184-8090265 (+) Ser (GCT) 82 bp Sc: 88.1\n",
      "-\n",
      "Homo_sapiens_tRNA-Val-CAC-1-1 (tRNAscan-SE ID: chr1.trna78) chr1:161369490-161369562 (-) Val (CAC) 73 bp Sc: 78.8;Homo_sapiens_tRNA-Val-CAC-1-2 (tRNAscan-SE ID: chr5.trna5) chr5:180524070-180524142 (+) Val (CAC) 73 bp Sc: 78.8;Homo_sapiens_tRNA-Val-CAC-1-3 (tRNAscan-SE ID: chr5.trna21) chr5:180529253-180529325 (-) Val (CAC) 73 bp Sc: 78.8;Homo_sapiens_tRNA-Val-CAC-1-4 (tRNAscan-SE ID: chr5.trna9) chr5:180600650-180600722 (+) Val (CAC) 73 bp Sc: 78.8;Homo_sapiens_tRNA-Val-CAC-1-5 (tRNAscan-SE ID: chr5.trna13) chr5:180649395-180649467 (-) Val (CAC) 73 bp Sc: 78.8;Homo_sapiens_tRNA-Val-CAC-1-6 (tRNAscan-SE ID: chr6.trna9) chr6:26538282-26538354 (+) Val (CAC) 73 bp Sc: 78.8\n",
      "-\n",
      "Homo_sapiens_tRNA-Val-TAC-1-1 (tRNAscan-SE ID: chr11.trna19) chr11:59318102-59318174 (-) Val (TAC) 73 bp Sc: 78.8;Homo_sapiens_tRNA-Val-TAC-1-2 (tRNAscan-SE ID: chrX.trna4) chrX:18693029-18693101 (-) Val (TAC) 73 bp Sc: 78.8\n",
      "-\n",
      "SNORD22;snoID_0137\n",
      "-\n"
     ]
    }
   ],
   "source": [
    "for row in df[\"Gene\"]:\n",
    "    if \";\" in row:\n",
    "        print(row)\n",
    "        print(\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb942324",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1f6acbce",
   "metadata": {},
   "source": [
    "# try m5C\n",
    "\n",
    "## columns that I need for data\n",
    "    chrom       \"PositionC\"[0]\n",
    "    chromStart  \"PositionC\"[1]\n",
    "    chromEnd    \n",
    "    name        m3C\n",
    "    score       BS_FDR\n",
    "    strand      Strand\n",
    "    thickStart   \n",
    "    thickEnd    \n",
    "    itemRGB     figure it out\n",
    "    coverage    BS_Depth\n",
    "    frequency   BS_%_non-converted\n",
    "    \n",
    "## information that I need for header\n",
    "    organism human\n",
    "    modification_type RNA\n",
    "    assembly GRCh37\n",
    "    annotation_source None\n",
    "    annotation_version None\n",
    "    sequencing_platform Illumina HiSeq 2500 (Illumina MiSeq for Psi data!!!!!)\n",
    "    bioinformatics_workflow https://github.com/HuntsmanCancerInstitute/RBSSeqTools/\n",
    "    experiment https://doi.org/10.1073/pnas.1817334116\n",
    "    external_source SRA:SRP093982\n",
    "    methods RBS-seq\n",
    "    references pubmed_id:30872485\n",
    "    conversion_information "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5aaab7e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Table Legends & Header Info', '1.High-threshold m5C candidates', '2.All m5C candidate sites', '3. RBS-Seq m5C Intersection'])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keys = parse_excel(\"/home/annebusch/anne02/euf-data/rbs-seq/GSE90963_Table_S1-m5C_candidate_sites.xlsx\")\n",
    "keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f241fe70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PositionC</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Classification</th>\n",
       "      <th>mRNA location</th>\n",
       "      <th>BS_Depth</th>\n",
       "      <th>BS_C+T</th>\n",
       "      <th>BS_C</th>\n",
       "      <th>BS_T</th>\n",
       "      <th>...</th>\n",
       "      <th>NBS_C+T</th>\n",
       "      <th>NBS_C</th>\n",
       "      <th>NBS_T</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster Nonconverted</th>\n",
       "      <th>Reverse-complement</th>\n",
       "      <th>Reverse-complement Nonconverted</th>\n",
       "      <th>Splice</th>\n",
       "      <th>Other</th>\n",
       "      <th>Passes Thresholds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:1277769</td>\n",
       "      <td>-</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>88</td>\n",
       "      <td>31</td>\n",
       "      <td>57</td>\n",
       "      <td>88</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:150530042</td>\n",
       "      <td>+</td>\n",
       "      <td>ADAMTSL4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>120</td>\n",
       "      <td>29</td>\n",
       "      <td>91</td>\n",
       "      <td>120</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:153665882</td>\n",
       "      <td>+</td>\n",
       "      <td>NPR1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:155030158</td>\n",
       "      <td>+</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>192</td>\n",
       "      <td>66</td>\n",
       "      <td>126</td>\n",
       "      <td>192</td>\n",
       "      <td>...</td>\n",
       "      <td>148</td>\n",
       "      <td>148</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:155158511</td>\n",
       "      <td>-</td>\n",
       "      <td>MUC1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>98</td>\n",
       "      <td>92</td>\n",
       "      <td>5</td>\n",
       "      <td>97</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>tRNA_60:47</td>\n",
       "      <td>+</td>\n",
       "      <td>Homo_sapiens_tRNA-Ala-TGC-7-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>3208</td>\n",
       "      <td>3207</td>\n",
       "      <td>2895</td>\n",
       "      <td>312</td>\n",
       "      <td>...</td>\n",
       "      <td>132</td>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>tRNA_63:49</td>\n",
       "      <td>+</td>\n",
       "      <td>Homo_sapiens_tRNA-Arg-ACG-1-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>1004</td>\n",
       "      <td>1004</td>\n",
       "      <td>452</td>\n",
       "      <td>552</td>\n",
       "      <td>...</td>\n",
       "      <td>2936</td>\n",
       "      <td>2935</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>tRNA_64:49</td>\n",
       "      <td>+</td>\n",
       "      <td>Homo_sapiens_tRNA-Arg-ACG-2-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>859</td>\n",
       "      <td>859</td>\n",
       "      <td>312</td>\n",
       "      <td>547</td>\n",
       "      <td>...</td>\n",
       "      <td>22460</td>\n",
       "      <td>22453</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>tRNA_78:61</td>\n",
       "      <td>+</td>\n",
       "      <td>Homo_sapiens_tRNA-Arg-TCT-1-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>224</td>\n",
       "      <td>224</td>\n",
       "      <td>98</td>\n",
       "      <td>126</td>\n",
       "      <td>...</td>\n",
       "      <td>7166</td>\n",
       "      <td>7164</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>tRNA_82:50</td>\n",
       "      <td>+</td>\n",
       "      <td>Homo_sapiens_tRNA-Arg-TCT-4-1 (tRNAscan-SE ID:...</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>88</td>\n",
       "      <td>81</td>\n",
       "      <td>72</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>486 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          PositionC Strand                                               Gene  \\\n",
       "0      chr1:1277769      -                                               DVL1   \n",
       "1    chr1:150530042      +                                           ADAMTSL4   \n",
       "2    chr1:153665882      +                                               NPR1   \n",
       "3    chr1:155030158      +                                             ADAM15   \n",
       "4    chr1:155158511      -                                               MUC1   \n",
       "..              ...    ...                                                ...   \n",
       "481      tRNA_60:47      +  Homo_sapiens_tRNA-Ala-TGC-7-1 (tRNAscan-SE ID:...   \n",
       "482      tRNA_63:49      +  Homo_sapiens_tRNA-Arg-ACG-1-1 (tRNAscan-SE ID:...   \n",
       "483      tRNA_64:49      +  Homo_sapiens_tRNA-Arg-ACG-2-1 (tRNAscan-SE ID:...   \n",
       "484      tRNA_78:61      +  Homo_sapiens_tRNA-Arg-TCT-1-1 (tRNAscan-SE ID:...   \n",
       "485      tRNA_82:50      +  Homo_sapiens_tRNA-Arg-TCT-4-1 (tRNAscan-SE ID:...   \n",
       "\n",
       "            Biotype  Classification mRNA location  BS_Depth  BS_C+T  BS_C  \\\n",
       "0    protein_coding  protein_coding        exonic        88      31    57   \n",
       "1    protein_coding  protein_coding        exonic       120      29    91   \n",
       "2    protein_coding  protein_coding        exonic        22      14     8   \n",
       "3    protein_coding  protein_coding        exonic       192      66   126   \n",
       "4    protein_coding  protein_coding          UTR3        98      92     5   \n",
       "..              ...             ...           ...       ...     ...   ...   \n",
       "481            tRNA            tRNA             .      3208    3207  2895   \n",
       "482            tRNA            tRNA             .      1004    1004   452   \n",
       "483            tRNA            tRNA             .       859     859   312   \n",
       "484            tRNA            tRNA             .       224     224    98   \n",
       "485            tRNA            tRNA             .        88      81    72   \n",
       "\n",
       "     BS_T  ...  NBS_C+T  NBS_C  NBS_T  Cluster  Cluster Nonconverted  \\\n",
       "0      88  ...       58     58      0        0                     0   \n",
       "1     120  ...       37     37      0        0                     0   \n",
       "2      22  ...       60     60      0        0                     0   \n",
       "3     192  ...      148    148      0        0                     0   \n",
       "4      97  ...       23     23      0        0                     0   \n",
       "..    ...  ...      ...    ...    ...      ...                   ...   \n",
       "481   312  ...      132    132      0        0                     0   \n",
       "482   552  ...     2936   2935      1        0                     0   \n",
       "483   547  ...    22460  22453      7        0                     0   \n",
       "484   126  ...     7166   7164      2        1                     0   \n",
       "485     9  ...      438    438      0        1                     0   \n",
       "\n",
       "     Reverse-complement  Reverse-complement Nonconverted  Splice  Other  \\\n",
       "0                     0                                0       1      0   \n",
       "1                     0                                0       1      0   \n",
       "2                     0                                0       0      0   \n",
       "3                     0                                0       0      0   \n",
       "4                     0                                0       0      0   \n",
       "..                  ...                              ...     ...    ...   \n",
       "481                   0                                0       0      0   \n",
       "482                   0                                0       0      0   \n",
       "483                   0                                0       0      0   \n",
       "484                   0                                0       0      0   \n",
       "485                   0                                0       0      0   \n",
       "\n",
       "     Passes Thresholds  \n",
       "0                    1  \n",
       "1                    1  \n",
       "2                    1  \n",
       "3                    1  \n",
       "4                    1  \n",
       "..                 ...  \n",
       "481                  1  \n",
       "482                  1  \n",
       "483                  1  \n",
       "484                  1  \n",
       "485                  1  \n",
       "\n",
       "[486 rows x 29 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel(\"/home/annebusch/anne02/euf-data/rbs-seq/GSE90963_Table_S1-m5C_candidate_sites.xlsx\", sheet_name=\"1.High-threshold m5C candidates\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "51cc9f9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = df[\"PositionC\"].str.startswith(\"chr\")\n",
    "\n",
    "df_chrom = df[mask]\n",
    "df_non_chrom = df[~mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f3946ec2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PositionC</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Classification</th>\n",
       "      <th>mRNA location</th>\n",
       "      <th>BS_Depth</th>\n",
       "      <th>BS_C+T</th>\n",
       "      <th>BS_C</th>\n",
       "      <th>BS_T</th>\n",
       "      <th>...</th>\n",
       "      <th>NBS_C+T</th>\n",
       "      <th>NBS_C</th>\n",
       "      <th>NBS_T</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster Nonconverted</th>\n",
       "      <th>Reverse-complement</th>\n",
       "      <th>Reverse-complement Nonconverted</th>\n",
       "      <th>Splice</th>\n",
       "      <th>Other</th>\n",
       "      <th>Passes Thresholds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:1277769</td>\n",
       "      <td>-</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>88</td>\n",
       "      <td>31</td>\n",
       "      <td>57</td>\n",
       "      <td>88</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:150530042</td>\n",
       "      <td>+</td>\n",
       "      <td>ADAMTSL4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>120</td>\n",
       "      <td>29</td>\n",
       "      <td>91</td>\n",
       "      <td>120</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:153665882</td>\n",
       "      <td>+</td>\n",
       "      <td>NPR1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:155030158</td>\n",
       "      <td>+</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>exonic</td>\n",
       "      <td>192</td>\n",
       "      <td>66</td>\n",
       "      <td>126</td>\n",
       "      <td>192</td>\n",
       "      <td>...</td>\n",
       "      <td>148</td>\n",
       "      <td>148</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:155158511</td>\n",
       "      <td>-</td>\n",
       "      <td>MUC1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>98</td>\n",
       "      <td>92</td>\n",
       "      <td>5</td>\n",
       "      <td>97</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>chrX:24095382</td>\n",
       "      <td>+</td>\n",
       "      <td>AluSx,EIF2S3</td>\n",
       "      <td>Repbase,protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>267</td>\n",
       "      <td>101</td>\n",
       "      <td>166</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>chrX:55206634</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>304</td>\n",
       "      <td>105</td>\n",
       "      <td>199</td>\n",
       "      <td>304</td>\n",
       "      <td>...</td>\n",
       "      <td>5385</td>\n",
       "      <td>5383</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>chrX:86958997</td>\n",
       "      <td>-</td>\n",
       "      <td>RPSAP15</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>1162</td>\n",
       "      <td>234</td>\n",
       "      <td>928</td>\n",
       "      <td>1162</td>\n",
       "      <td>...</td>\n",
       "      <td>130</td>\n",
       "      <td>130</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>chrY:20777415</td>\n",
       "      <td>-</td>\n",
       "      <td>HUERS-P3b-int</td>\n",
       "      <td>Repbase</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>44</td>\n",
       "      <td>28</td>\n",
       "      <td>16</td>\n",
       "      <td>44</td>\n",
       "      <td>...</td>\n",
       "      <td>936</td>\n",
       "      <td>936</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>chrY:20866759</td>\n",
       "      <td>+</td>\n",
       "      <td>HUERS-P3b-int</td>\n",
       "      <td>Repbase</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>.</td>\n",
       "      <td>36</td>\n",
       "      <td>26</td>\n",
       "      <td>10</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>937</td>\n",
       "      <td>937</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>189 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          PositionC Strand           Gene                 Biotype  \\\n",
       "0      chr1:1277769      -           DVL1          protein_coding   \n",
       "1    chr1:150530042      +       ADAMTSL4          protein_coding   \n",
       "2    chr1:153665882      +           NPR1          protein_coding   \n",
       "3    chr1:155030158      +         ADAM15          protein_coding   \n",
       "4    chr1:155158511      -           MUC1          protein_coding   \n",
       "..              ...    ...            ...                     ...   \n",
       "184   chrX:24095382      +   AluSx,EIF2S3  Repbase,protein_coding   \n",
       "185   chrX:55206634      -              .                       .   \n",
       "186   chrX:86958997      -        RPSAP15              pseudogene   \n",
       "187   chrY:20777415      -  HUERS-P3b-int                 Repbase   \n",
       "188   chrY:20866759      +  HUERS-P3b-int                 Repbase   \n",
       "\n",
       "     Classification mRNA location  BS_Depth  BS_C+T  BS_C  BS_T  ...  NBS_C+T  \\\n",
       "0    protein_coding        exonic        88      31    57    88  ...       58   \n",
       "1    protein_coding        exonic       120      29    91   120  ...       37   \n",
       "2    protein_coding        exonic        22      14     8    22  ...       60   \n",
       "3    protein_coding        exonic       192      66   126   192  ...      148   \n",
       "4    protein_coding          UTR3        98      92     5    97  ...       23   \n",
       "..              ...           ...       ...     ...   ...   ...  ...      ...   \n",
       "184  protein_coding          UTR3       267     101   166   267  ...      119   \n",
       "185           ncRNA             .       304     105   199   304  ...     5385   \n",
       "186      pseudogene             .      1162     234   928  1162  ...      130   \n",
       "187           ncRNA             .        44      28    16    44  ...      936   \n",
       "188           ncRNA             .        36      26    10    36  ...      937   \n",
       "\n",
       "     NBS_C  NBS_T  Cluster  Cluster Nonconverted  Reverse-complement  \\\n",
       "0       58      0        0                     0                   0   \n",
       "1       37      0        0                     0                   0   \n",
       "2       60      0        0                     0                   0   \n",
       "3      148      0        0                     0                   0   \n",
       "4       23      0        0                     0                   0   \n",
       "..     ...    ...      ...                   ...                 ...   \n",
       "184    119      0        0                     0                   0   \n",
       "185   5383      2        1                     0                   0   \n",
       "186    130      0        0                     0                   0   \n",
       "187    936      0        1                     0                   0   \n",
       "188    937      0        1                     0                   0   \n",
       "\n",
       "     Reverse-complement Nonconverted  Splice  Other  Passes Thresholds  \n",
       "0                                  0       1      0                  1  \n",
       "1                                  0       1      0                  1  \n",
       "2                                  0       0      0                  1  \n",
       "3                                  0       0      0                  1  \n",
       "4                                  0       0      0                  1  \n",
       "..                               ...     ...    ...                ...  \n",
       "184                                0       0      0                  1  \n",
       "185                                0       0      0                  1  \n",
       "186                                0       0      0                  1  \n",
       "187                                0       0      0                  1  \n",
       "188                                0       0      0                  1  \n",
       "\n",
       "[189 rows x 29 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "157ea912",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>trna</th>\n",
       "      <th>strand</th>\n",
       "      <th>num_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>offset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7990338</td>\n",
       "      <td>7990408</td>\n",
       "      <td>tRNA-Und-NNN-4-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16847079</td>\n",
       "      <td>16847153</td>\n",
       "      <td>tRNA-Asn-GTT-5-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16858892</td>\n",
       "      <td>16858966</td>\n",
       "      <td>tRNA-Asn-GTT-18-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16861773</td>\n",
       "      <td>16861845</td>\n",
       "      <td>tRNA-Glu-TTC-4-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16872433</td>\n",
       "      <td>16872504</td>\n",
       "      <td>tRNA-Gly-CCC-1-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>chrX</td>\n",
       "      <td>3833270</td>\n",
       "      <td>3833344</td>\n",
       "      <td>tRNA-Ile-GAT-1-3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>chrX</td>\n",
       "      <td>3833684</td>\n",
       "      <td>3833758</td>\n",
       "      <td>tRX-Ile-NNN-1-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18693028</td>\n",
       "      <td>18693101</td>\n",
       "      <td>tRNA-Val-TAC-1-2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25280088</td>\n",
       "      <td>25280160</td>\n",
       "      <td>tRX-Ser-NNN-2-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>chrX</td>\n",
       "      <td>118737860</td>\n",
       "      <td>118737929</td>\n",
       "      <td>tRX-Pro-NNN-1-1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end               trna strand  num_exons  length  \\\n",
       "0    chr1    7990338    7990408   tRNA-Und-NNN-4-1      -          1      70   \n",
       "1    chr1   16847079   16847153   tRNA-Asn-GTT-5-1      -          1      74   \n",
       "2    chr1   16858892   16858966  tRNA-Asn-GTT-18-1      -          1      74   \n",
       "3    chr1   16861773   16861845   tRNA-Glu-TTC-4-1      -          1      72   \n",
       "4    chr1   16872433   16872504   tRNA-Gly-CCC-1-1      -          1      71   \n",
       "..    ...        ...        ...                ...    ...        ...     ...   \n",
       "591  chrX    3833270    3833344   tRNA-Ile-GAT-1-3      -          1      74   \n",
       "592  chrX    3833684    3833758    tRX-Ile-NNN-1-1      -          1      74   \n",
       "593  chrX   18693028   18693101   tRNA-Val-TAC-1-2      -          1      73   \n",
       "594  chrX   25280088   25280160    tRX-Ser-NNN-2-1      -          1      72   \n",
       "595  chrX  118737860  118737929    tRX-Pro-NNN-1-1      +          1      69   \n",
       "\n",
       "     offset  \n",
       "0         0  \n",
       "1         0  \n",
       "2         0  \n",
       "3         0  \n",
       "4         0  \n",
       "..      ...  \n",
       "591       0  \n",
       "592       0  \n",
       "593       0  \n",
       "594       0  \n",
       "595       0  \n",
       "\n",
       "[596 rows x 8 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_non_chrom = df_non_chrom.reset_index(drop=True)\n",
    "df_non_chrom\n",
    "tRNA = pd.read_csv(\"/home/annebusch/anne02/tRNA-index/human/hg19-tRNAs.bed\", delimiter=\"\\t\", names = [\"chrom\", \"start\", \"end\", \"trna\", \"score\", \"strand\", \"thickstart\", \"thickend\", \"zero\", \"num_exons\", \"length\", \"offset\"], header=None)\n",
    "del tRNA[\"zero\"]\n",
    "del tRNA[\"thickstart\"]\n",
    "del tRNA[\"thickend\"]\n",
    "del tRNA[\"score\"]\n",
    "tRNA['length'] = tRNA['length'].str.replace(',', '').astype(int)\n",
    "tRNA['offset'] = tRNA['offset'].str.replace(',', '').astype(int)\n",
    "tRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e674259b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['PositionC', 'Strand', 'Gene', 'Biotype', 'Classification',\n",
       "       'mRNA location', 'BS_Depth', 'BS_C+T', 'BS_C', 'BS_T',\n",
       "       'BS_%_non-converted', 'BS_FDR', 'BS_Depth.1', 'BS_C+T.1', 'BS_C.1',\n",
       "       'BS_T.1', 'BS_%_non-converted.1', 'BS_FDR.1', 'NBS_Depth', 'NBS_C+T',\n",
       "       'NBS_C', 'NBS_T', 'Cluster', 'Cluster Nonconverted',\n",
       "       'Reverse-complement', 'Reverse-complement Nonconverted', 'Splice',\n",
       "       'Other', 'Passes Thresholds'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chrom.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea9a13a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d14be688",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>trna</th>\n",
       "      <th>strand</th>\n",
       "      <th>num_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>offset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7990338</td>\n",
       "      <td>7990408</td>\n",
       "      <td>tRNA-Und-NNN-4-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16847079</td>\n",
       "      <td>16847153</td>\n",
       "      <td>tRNA-Asn-GTT-5-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16858892</td>\n",
       "      <td>16858966</td>\n",
       "      <td>tRNA-Asn-GTT-18-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16861773</td>\n",
       "      <td>16861845</td>\n",
       "      <td>tRNA-Glu-TTC-4-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16872433</td>\n",
       "      <td>16872504</td>\n",
       "      <td>tRNA-Gly-CCC-1-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>chrX</td>\n",
       "      <td>3833270</td>\n",
       "      <td>3833344</td>\n",
       "      <td>tRNA-Ile-GAT-1-3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>chrX</td>\n",
       "      <td>3833684</td>\n",
       "      <td>3833758</td>\n",
       "      <td>tRX-Ile-NNN-1-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18693028</td>\n",
       "      <td>18693101</td>\n",
       "      <td>tRNA-Val-TAC-1-2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25280088</td>\n",
       "      <td>25280160</td>\n",
       "      <td>tRX-Ser-NNN-2-1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>chrX</td>\n",
       "      <td>118737860</td>\n",
       "      <td>118737929</td>\n",
       "      <td>tRX-Pro-NNN-1-1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end               trna strand  num_exons  length  \\\n",
       "0    chr1    7990338    7990408   tRNA-Und-NNN-4-1      -          1      70   \n",
       "1    chr1   16847079   16847153   tRNA-Asn-GTT-5-1      -          1      74   \n",
       "2    chr1   16858892   16858966  tRNA-Asn-GTT-18-1      -          1      74   \n",
       "3    chr1   16861773   16861845   tRNA-Glu-TTC-4-1      -          1      72   \n",
       "4    chr1   16872433   16872504   tRNA-Gly-CCC-1-1      -          1      71   \n",
       "..    ...        ...        ...                ...    ...        ...     ...   \n",
       "591  chrX    3833270    3833344   tRNA-Ile-GAT-1-3      -          1      74   \n",
       "592  chrX    3833684    3833758    tRX-Ile-NNN-1-1      -          1      74   \n",
       "593  chrX   18693028   18693101   tRNA-Val-TAC-1-2      -          1      73   \n",
       "594  chrX   25280088   25280160    tRX-Ser-NNN-2-1      -          1      72   \n",
       "595  chrX  118737860  118737929    tRX-Pro-NNN-1-1      +          1      69   \n",
       "\n",
       "     offset  \n",
       "0         0  \n",
       "1         0  \n",
       "2         0  \n",
       "3         0  \n",
       "4         0  \n",
       "..      ...  \n",
       "591       0  \n",
       "592       0  \n",
       "593       0  \n",
       "594       0  \n",
       "595       0  \n",
       "\n",
       "[596 rows x 8 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f879a3ec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6617e01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0a9b355",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b350458",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
